### PR TITLE
CASMPET-6398: kyverno prevents weave-net daemonset from creating pods

### PR DESCRIPTION
**Pull request by pradeepkumargl**
## Summary and Scope

The daemonset-controller is not able to create the weave pod because it cannot get to kyverno to call the webhook.

## Testing

Before Change
-------------

```
pit:~ # kubectl get pods -o wide -A | grep weave
kube-system      weave-net-45np9                                                   2/2     Running            0          2m21s   10.252.1.9    ncn-w003   <none>           <none>
kube-system      weave-net-67ttc                                                   2/2     Running            1          2d16h   10.252.1.8    ncn-w002   <none>           <none>
kube-system      weave-net-92tls                                                   0/2     Terminating        1          2d16h   10.252.1.11   ncn-m002   <none>           <none>
kube-system      weave-net-g9d5f                                                   2/2     Running            0          77s     10.252.1.7    ncn-w001   <none>           <none>
kube-system      weave-net-nfwxb                                                   2/2     Running            0          3m13s   10.252.1.12   ncn-m003   <none>           <none>
pit:~ # journalctl -u kubelet | grep webhook
.
.
.
Mar 15 07:45:12 ncn-m002 kubelet[14609]: I0315 07:45:12.693060   14609 status_manager.go:611] "Failed to delete status for pod" pod="kube-system/weave-net-9r5x2" err="Internal error occurred: failed calling webhook \"mutate.kyverno.svc-fail\": Post \"https://cray-kyverno-svc.kyverno.svc:443/mutate?timeout=10s\": dial tcp 10.26.61.128:443: connect: no route to host"
Mar 15 07:45:22 ncn-m002 kubelet[14609]: I0315 07:45:22.704503   14609 status_manager.go:611] "Failed to delete status for pod" pod="kube-system/weave-net-9r5x2" err="Internal error occurred: failed calling webhook \"mutate.kyverno.svc-fail\": Post \"https://cray-kyverno-svc.kyverno.svc:443/mutate?timeout=10s\": dial tcp 10.26.61.128:443: connect: no route to host"
Mar 15 07:45:32 ncn-m002 kubelet[14609]: I0315 07:45:32.688340   14609 status_manager.go:611] "Failed to delete status for pod" pod="kube-system/weave-net-9r5x2" err="Internal error occurred: failed calling webhook \"mutate.kyverno.svc-fail\": Post \"https://cray-kyverno-svc.kyverno.svc:443/mutate?timeout=10s\": dial tcp 10.26.61.128:443: connect: no route to host"
Mar 15 07:45:42 ncn-m002 kubelet[14609]: I0315 07:45:42.708703   14609 status_manager.go:611] "Failed to delete status for pod" pod="kube-system/weave-net-9r5x2" err="Internal error occurred: failed calling webhook \"mutate.kyverno.svc-fail\": Post \"https://cray-kyverno-svc.kyverno.svc:443/mutate?timeout=10s\": dial tcp 10.26.61.128:443: connect: no route to host"
Mar 15 07:45:52 ncn-m002 kubelet[14609]: I0315 07:45:52.692738   14609 status_manager.go:611] "Failed to delete status for pod" pod="kube-system/weave-net-9r5x2" err="Internal error occurred: failed calling webhook \"mutate.kyverno.svc-fail\": Post \"https://cray-kyverno-svc.kyverno.svc:443/mutate?timeout=10s\": dial tcp 10.26.61.128:443: connect: no route to host"
ncn-m002:~ # kubectl get pods -o wide -A | grep weave
kube-system      weave-net-6dqds                                                   2/2     Running            0          45m     10.252.1.8    ncn-w002   <none>           <none>
kube-system      weave-net-9r5x2                                                   0/2     Terminating        0          44m     10.252.1.11   ncn-m002   <none>           <none>
kube-system      weave-net-ct5mg                                                   2/2     Running            0          37m     10.252.1.12   ncn-m003   <none>           <none>
kube-system      weave-net-sbh22                                                   2/2     Running            0          38m     10.252.1.7    ncn-w001   <none>           <none>
kube-system      weave-net-vdnk8                                                   2/2     Running            0          46m     10.252.1.9    ncn-w003   <none>           <none>
ncn-m002:~ #
```

After change
------------

```
pit:~ # kubectl edit configmap cray-kyverno -n kyverno
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
data:
  generateSuccessEvents: "false"
  resourceFilters: '[Event,*,*][*,kube-system,*][*,kube-public,*][*,kube-node-lease,*][Node,*,*][APIService,*,*][TokenReview,*,*][SubjectAccessReview,*,*][SelfSubjectAccessReview,*,*][*,kyverno,kyverno*][Binding,*,*][ReplicaSet,*,*][ReportChangeRequest,*,*][ClusterReportChangeRequest,*,*]'
  webhooks: '[{"namespaceSelector":{"matchExpressions":[{"key":"kubernetes.io/metadata.name","operator":"NotIn","values":["kube-system"]}]}}]'
kind: ConfigMap
metadata:
  annotations:
    meta.helm.sh/release-name: cray-kyverno
    meta.helm.sh/release-namespace: kyverno
  creationTimestamp: "2023-03-12T11:37:44Z"
  labels:
    app: kyverno
    app.kubernetes.io/component: kyverno
    app.kubernetes.io/instance: cray-kyverno
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kyverno
    app.kubernetes.io/part-of: kyverno
    app.kubernetes.io/version: v2.3.3
    helm.sh/chart: kyverno-v2.3.3
  name: cray-kyverno
  namespace: kyverno
  resourceVersion: "3968894"
  uid: 96549bfb-fbe1-4781-a25f-a39b69dfe0b0
pit:~ # kubectl get cm cray-kyverno -n kyverno -o yaml
apiVersion: v1
data:
  generateSuccessEvents: "false"
  resourceFilters: '[Event,*,*][*,kube-system,*][*,kube-public,*][*,kube-node-lease,*][Node,*,*][APIService,*,*][TokenReview,*,*][SubjectAccessReview,*,*][SelfSubjectAccessReview,*,*][*,kyverno,kyverno*][Binding,*,*][ReplicaSet,*,*][ReportChangeRequest,*,*][ClusterReportChangeRequest,*,*]'
  webhooks: '[{"namespaceSelector":{"matchExpressions":[{"key":"kubernetes.io/metadata.name","operator":"NotIn","values":["kube-system"]}]}}]'
kind: ConfigMap
metadata:
  annotations:
    meta.helm.sh/release-name: cray-kyverno
    meta.helm.sh/release-namespace: kyverno
  creationTimestamp: "2023-03-12T11:37:44Z"
  labels:
    app: kyverno
    app.kubernetes.io/component: kyverno
    app.kubernetes.io/instance: cray-kyverno
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kyverno
    app.kubernetes.io/part-of: kyverno
    app.kubernetes.io/version: v2.3.3
    helm.sh/chart: kyverno-v2.3.3
  name: cray-kyverno
  namespace: kyverno
  resourceVersion: "4011782"
  uid: 96549bfb-fbe1-4781-a25f-a39b69dfe0b0
pit:~ #
pit:~ # kubectl rollout restart daemonset weave-net -n kube-system
daemonset.apps/weave-net restarted
pit:~ #
pit:~ # kubectl get pods -o wide -A | grep weave
kube-system      weave-net-4m8nx                                                   2/2     Running            0          100s    10.252.1.11   ncn-m002   <none>           <none>
kube-system      weave-net-58mqr                                                   2/2     Running            0          2m25s   10.252.1.7    ncn-w001   <none>           <none>
kube-system      weave-net-874bd                                                   2/2     Running            0          3m7s    10.252.1.8    ncn-w002   <none>           <none>
kube-system      weave-net-8stwz                                                   2/2     Running            0          3m53s   10.252.1.9    ncn-w003   <none>           <none>
kube-system      weave-net-knxcq                                                   2/2     Running            0          53s     10.252.1.12   ncn-m003   <none>           <none>
pit:~ # journalctl -u kubelet | grep webhook
pit:~ #
```

### Tested on:

Dorian

### CODE REVIEW

File: charts/kyverno/values.yaml [hunk 1]
```diff
@@ -38,3 +38,5 @@ kyverno:
   priorityClassName: csm-high-priority-service
  # Desired number of pods/replica
   replicaCount: 3
+  config:
+    webhooks: [{"namespaceSelector":{"matchExpressions":[{"key":"kubernetes.io/metadata.name","operator":"NotIn","values":["kube-system"]}]}}]
```
Comment (johren-hpe) [inline charts/kyverno/values.yaml:42]: @jeremy-duckworth mentioned in the JIRA ticket that we need to "ensure that the kyverno namespace exclude is not overridden".   I think we also need to add that namespace to the webhook namespaceSelector configuration.

https://github.com/kyverno/kyverno/blob/main/charts/kyverno/templates/config/configmap.yaml#L35-L41

**Review by johren-hpe — CHANGES_REQUESTED** [review]

**Comment by johren-hpe** [comment]
I'll also note that this repo currently doesn't require any reviews to merge.   Is that expected?

**Comment by johren-hpe** [comment]
It looks like this repo also doesn't have an action to build the chart.   That would help to check that there are no build issues before merging and it would also provide a chart that could be tested.

**Comment by johren-hpe** [comment]
I'm guessing the actions are not building the chart because this PR is from a branch on a fork.

File: charts/kyverno/values.yaml (hunk 1 above)
Comment (johren-hpe) [reply charts/kyverno/values.yaml:42]: I forgot that we are using kyverno 1.6.2 which is several version off of main.   1.6.2 doesn't use `excludeKyvernoNamespace` value in configmap.yaml and doesn't define it in values.yaml.

https://github.com/kyverno/kyverno/blob/v1.6.2/charts/kyverno/templates/configmap.yaml#L20-L22

So I think that means that the kyverno exclude needs to be specified in `config.webhooks` the same way it is for the `kube-system` exclude.

Comment (yarlagadda-srinivasa-rao) [reply charts/kyverno/values.yaml:42]: Yes @johren-hpe, Thank you. @pradeepkumargl missed to add this. Hope changing it as below to exclude both.
webhooks: [{"namespaceSelector":{"matchExpressions":[{"key":"kubernetes.io/metadata.name","operator":"NotIn","values":["kyverno","kube-system"]}]}}]

Comment (yarlagadda-srinivasa-rao) [reply charts/kyverno/values.yaml:42]: OR, We need add something like for better readability?
config:
  webhooks:
  - namespaceSelector:
      matchExpressions:
      - key: kubernetes.io/metadata.name
        operator: NotIn
        values:
        - kyverno
        - kube-system

**Comment by yarlagadda-srinivasa-rao** [comment]
Manual chart creation and installation logs we collected before raising the PR will attach those logs for the reference.

**Comment by pradeepkumargl** [comment]
Package creation and installation logs
-----------------------------------------
pit:/tmp/pradeep/cray-kyverno # vi values.yaml
pit:/tmp/pradeep/cray-kyverno # cd ..
pit:/tmp/pradeep # helm package cray-kyverno
Successfully packaged chart and saved it to: /tmp/pradeep/cray-kyverno-1.4.0.tgz
pit:/tmp/pradeep # ls -ltr
total 44
-rw-r--r-- 1 root root   156 Mar 15 17:56 nginx.yaml
drwxr-xr-x 4 root root   128 Mar 15 23:44 cray-kyverno
-rw-r--r-- 1 root root 38769 Mar 15 23:44 cray-kyverno-1.4.0.tgz
pit:/tmp/pradeep # helm upgrade -n kyverno cray-kyverno /tmp/pradeep/cray-kyverno-1.4.0.tgz
Release "cray-kyverno" has been upgraded. Happy Helming!
NAME: cray-kyverno
LAST DEPLOYED: Wed Mar 15 23:45:11 2023
NAMESPACE: kyverno
STATUS: deployed
REVISION: 6
NOTES:
The official kyverno chart installed with some cray-specific overrides and defaults
pit:/tmp/pradeep #
pit:/tmp/pradeep # kubectl get cm cray-kyverno -n kyverno -o yaml
apiVersion: v1
data:
  generateSuccessEvents: "false"
  resourceFilters: '[Event,*,*][*,kube-system,*][*,kube-public,*][*,kube-node-lease,*][Node,*,*][APIService,*,*][TokenReview,*,*][SubjectAccessReview,*,*][SelfSubjectAccessReview,*,*][*,kyverno,kyverno*][Binding,*,*][ReplicaSet,*,*][ReportChangeRequest,*,*][ClusterReportChangeRequest,*,*]'
  webhooks: '[{"namespaceSelector":{"matchExpressions":[{"key":"kubernetes.io/metadata.name","operator":"NotIn","values":["kyverno","kube-system"]}]}}]'
kind: ConfigMap
metadata:
  annotations:
    meta.helm.sh/release-name: cray-kyverno
    meta.helm.sh/release-namespace: kyverno
  creationTimestamp: "2023-03-12T11:37:44Z"
  labels:
    app: kyverno
    app.kubernetes.io/component: kyverno
    app.kubernetes.io/instance: cray-kyverno
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kyverno
    app.kubernetes.io/part-of: kyverno
    app.kubernetes.io/version: v2.3.3
    helm.sh/chart: kyverno-v2.3.3
  name: cray-kyverno
  namespace: kyverno
  resourceVersion: "4938066"
  uid: 96549bfb-fbe1-4781-a25f-a39b69dfe0b0
pit:/tmp/pradeep #
pit:/tmp/pradeep # kubectl get pods -o wide -A | grep weave
kube-system      weave-net-j9bzl                                                   1/2     Running     0          5s      10.252.1.8    ncn-w002   <none>           <none>
kube-system      weave-net-rlg4k                                                   2/2     Running     0          58s     10.252.1.12   ncn-m003   <none>           <none>
kube-system      weave-net-s45bh                                                   2/2     Running     0          105s    10.252.1.9    ncn-w003   <none>           <none>
kube-system      weave-net-vsjn9                                                   2/2     Running     0          3m14s   10.252.1.11   ncn-m002   <none>           <none>
kube-system      weave-net-xhgsx                                                   2/2     Running     0          2m27s   10.252.1.7    ncn-w001   <none>           <none>
pit:/tmp/pradeep # date
Wed Mar 15 23:56:21 UTC 2023
pit:/tmp/pradeep #

File: charts/kyverno/values.yaml (hunk 1 above)
Comment (pradeepkumargl) [reply charts/kyverno/values.yaml]: included kyverno along with kube-system and submitted the changes after testing.

**Review by johren-hpe — APPROVED** [review]

**Comment by jeremy-duckworth** [comment]
> I'll also note that this repo currently doesn't require any reviews to merge. Is that expected?

No. @yarlagadda-srinivasa-rao please require at least one approved reviewer from PET or SEC. If you need help setting policy, please advise but you should be authorized to make this change.

**Review by jeremy-duckworth — APPROVED** [review]

**Pull request merged**